Translate DNR initiatorDomains and excludedInitiatiorDomains into if-frame-url and unless-frame-url

b-weinstein opened the pull request:
#### 063bb419b220cafc9bb10b6e18f05468b9666607
<pre>
Translate DNR initiatorDomains and excludedInitiatiorDomains into if-frame-url and unless-frame-url
<a href="https://bugs.webkit.org/show_bug.cgi?id=282783">https://bugs.webkit.org/show_bug.cgi?id=282783</a>
<a href="https://rdar.apple.com/97014647">rdar://97014647</a>

Reviewed by Timothy Hatcher.

One thing that this PR doesn&apos;t handle correctly is if a DNR rule specifies both initiatorDomains
and excludedInitiatorDomains. In that case, we should be translating that rule into a block rule
and an ignore-previous-rules rule, but we aren&apos;t doing that yet.

* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.mm:
(WebKit::if):
(-[_WKWebExtensionDeclarativeNetRequestRule _webKitRuleWithWebKitActionType:chromeActionType:chromeResourceTypes:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIDeclarativeNetRequest.mm:
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, RuleConversionWithInitiatorDomains)):
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, RuleConversionWithExcludedInitiatorDomains)):

Canonical link: <a href="https://commits.webkit.org/286340@main">https://commits.webkit.org/286340@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fc51480b8066056e74f7af1ac80d374c39580dd3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75595 "Passed style check") | [⏳ 🛠 ios ](https://ews-build.webkit.org/#/builders/iOS-18-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28446 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80075 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26859 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77711 "Passed tests") | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2883 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59283 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17475 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78662 "Passed tests") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64931 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39644 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22416 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25188 "Built successfully") | 
| | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22753 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81554 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2934 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1849 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67523 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3085 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64907 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66823 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10787 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8935 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11682 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2891 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/5713 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2916 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3851 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2923 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIDeclarativeNetRequest.mm:
```diff
@@ -1958,6 +1958,66 @@
+        @"id": @1,
+        @"action": @{ @"type": @"block" },
+        @"condition": @{
+            // Some extensions will have added leading * to match subdomains to work around Safari's bug in the past (rdar://113865040), make sure we still handle that case correctly.
```
```suggestion
```

b-weinstein replied:
Oh whoops - I removed that comment during my development work, but I was editing the wrong test so I undid it and forgot to remove it again.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIDeclarativeNetRequest.mm:
```diff
@@ -1958,6 +1958,66 @@
+        @"id": @1,
+        @"action": @{ @"type": @"block" },
+        @"condition": @{
+            // Some extensions will have added leading * to match subdomains to work around Safari's bug in the past (rdar://113865040), make sure we still handle that case correctly.
```
```suggestion
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.mm:
```diff
@@ -686,7 +709,6 @@ - (void)removeInvalidResourceTypesForKey:(NSString *)ruleConditionKey
     }
 }
 
-// If you are making changes to how Chrome's rules are translated into WebKit rules, increase the hash version number in _WKWebExtensionDeclarativeNetRequestTranslator -- currentDeclarativeNetRequestHashVersion.
```
Keep?

b-weinstein replied:
This is a bit outdated - that variable doesn't exist anymore.

The correct comment is further up in the file:
```
// If any changes are made to the rule translation logic here, make sure to increment currentDeclarativeNetRequestRuleTranslatorVersion in WebExtensionContextCocoa.
```

I need to update that version as well.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.mm:
```diff
@@ -334,6 +338,25 @@ - (instancetype)initWithDictionary:(NSDictionary *)ruleDictionary errorString:(N
+    if (NSArray<NSString *> *excludedInitiatorDomains = _condition[ruleConditionExcludedInitiatorDomainsKey]) {
+        if (!isArrayOfDomainsValid(excludedInitiatorDomains)) {
+            if (outErrorString)
+                *outErrorString = [NSString stringWithFormat:@"Rule with id %ld is invalid. `excludedInitiatorDomains` must be non-empty and cannot contain non-ASCII characters.", (long)_ruleID];
```
Empty is allowed, per MDN. "If the list is empty or omitted, no domains are excluded."
```suggestion
        if (!isArrayOfExcludedDomainsValid(excludedInitiatorDomains)) {
            if (outErrorString)
                *outErrorString = [NSString stringWithFormat:@"Rule with id %ld is invalid. `excludedInitiatorDomains` cannot contain non-ASCII characters.", (long)_ruleID];
```

b-weinstein replied:
Whoops, missed the part in the error message here.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.mm:
```diff
@@ -334,6 +338,25 @@ - (instancetype)initWithDictionary:(NSDictionary *)ruleDictionary errorString:(N
+    if (NSArray<NSString *> *excludedInitiatorDomains = _condition[ruleConditionExcludedInitiatorDomainsKey]) {
+        if (!isArrayOfExcludedDomainsValid(excludedInitiatorDomains)) {
+            if (outErrorString)
+                *outErrorString = [NSString stringWithFormat:@"Rule with id %ld is invalid. `excludedInitiatorDomains` must be non-empty and cannot contain non-ASCII characters.", (long)_ruleID];
```
```suggestion
                *outErrorString = [NSString stringWithFormat:@"Rule with id %ld is invalid. `excludedInitiatorDomains` cannot contain non-ASCII characters.", (long)_ruleID];
```

b-weinstein committed:
Translate DNR initiatorDomains and excludedInitiatiorDomains into if-frame-url and unless-frame-url
https://bugs.webkit.org/show_bug.cgi?id=282783
rdar://97014647

Reviewed by Timothy Hatcher.

One thing that this PR doesn't handle correctly is if a DNR rule specifies both initiatorDomains
and excludedInitiatorDomains. In that case, we should be translating that rule into a block rule
and an ignore-previous-rules rule, but we aren't doing that yet.

* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.mm:
(WebKit::if):
(-[_WKWebExtensionDeclarativeNetRequestRule _webKitRuleWithWebKitActionType:chromeActionType:chromeResourceTypes:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIDeclarativeNetRequest.mm:
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, RuleConversionWithInitiatorDomains)):
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, RuleConversionWithExcludedInitiatorDomains)):

Canonical link: https://commits.webkit.org/286340@main
